### PR TITLE
Upgrade Elasticsearch from 1.4.5 to 5.x

### PR DESCRIPTION
**Pull request by yih-en**
### Notable Changes

- Usable in Elasticsearch 6.1.1 also
- Replace ```tire``` gem with ```elasticsearch-rails``` gem
- ```_timestamp``` is now deprecated in Elasticsearch 5.x. Replace with ```timestamp``` field instead
- ```_ttl``` is now deprecated in Elasticsearch 5.x. See alternatives [here](https://www.elastic.co/guide/en/elasticsearch/reference/2.0/mapping-ttl-field.html)

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -8,8 +8,8 @@ end
 gem 'rails', '5.1.4'
 gem 'rails-controller-testing', '~> 1.0'
 gem 'nokogiri', '~> 1.8.0'
-gem 'tire', '~> 0.6.2' #deprecated in 2013
-gem 'tire-contrib', '~> 0.1.2'
+# gem 'tire', '~> 0.6.2' #deprecated in 2013
```
Comment (nickmarden) [inline Gemfile]: If removing, please remove rather than commenting out.

File: config/application.rb [hunk 2]
```diff
@@ -14,6 +14,8 @@ class Application < Rails::Application
     # Configure the default encoding used in templates for Ruby 1.9.
     config.encoding = 'utf-8'
 
+    config.elasticsearch_config = (YAML.load_file("#{Rails.root}/config/elasticsearch.yml") || {})[Rails.env]
```
Comment (nickmarden) [inline config/application.rb]: I think `config_for(:elasticsearch)` does the same thing?

File: config/environments/development.rb [hunk 3]
```diff
@@ -32,4 +32,9 @@
   # Use an evented file watcher to asynchronously detect changes in source code,
   # routes, locales, etc. This feature depends on the listen gem.
   config.file_watcher = ActiveSupport::EventedFileUpdateChecker
+
+  config.elasticsearch_client = Elasticsearch::Client.new(
+    host: 'localhost:9200',
+    log: true
```
Comment (nickmarden) [inline config/environments/development.rb]: Need to support username/password since we are using a password-protected ES 5.x cluster.

File: app/models/position_opening.rb [hunk 4]
```diff
@@ -1,134 +1,238 @@
...
+      search_definition = {
+        query: {
+          bool: {
+            must: [],
+            should: [],
+            filter: [
+              { range: { start_date: { lte: Date.current } }}
+            ],
+            minimum_should_match: '0<1'
+          }
```
Comment (yih-en) [inline app/models/position_opening.rb]: This is related to https://github.com/elastic/elasticsearch/issues/21844

File: config/environments/production.rb [hunk 5]
```diff
@@ -49,4 +49,8 @@
     logger.formatter = config.log_formatter
     config.logger    = ActiveSupport::TaggedLogging.new(logger)
   end
+
+  config.elasticsearch_client = Elasticsearch::Client.new(
```
Comment (nickmarden) [inline config/environments/production.rb]: I feel that, rather than being hardcoded for the three different environments, this should be initialized in `config/environment.rb` based on the contents of the `elasticsearch.yml` file. Feel free to change the format of the `config/elasticsearch.yml` file as needed to make that work; we can adjust our recipes.

File: config/initializers/elasticsearch.rb [hunk 6]
```diff
@@ -1,7 +1,5 @@
 module Elasticsearch; end
 
-es_config = (YAML.load_file("#{Rails.root}/config/elasticsearch.yml") || {})[Rails.env]
-
-Tire::Configuration.url(es_config['url']) if es_config && es_config['url'].present?
+es_config = Rails.application.config.elasticsearch_config
```
Comment (nickmarden) [inline config/initializers/elasticsearch.rb]: I prefer `Rails.application.config.elasticsearch`, the second `_config` is redundant.

File: app/models/position_opening.rb [hunk 7]
```diff
@@ -1,134 +1,238 @@
...
+      if query.keywords.blank?
+        if lat.blank? || lon.blank?
+          search_definition[:sort] = [{ "#{sort_by}": 'desc' }]
+        else
+          search_definition[:sort] = [{
+            _geo_distance: {
+              'locations.geo': { lat: lat.to_f, lon: lon.to_f },
+              order: 'asc',
+              nested_path: 'locations'
+            }
```
Comment (yih-en) [inline app/models/position_opening.rb]: nested_path is required when sorting nested fields: https://www.elastic.co/guide/en/elasticsearch/reference/5.2/search-request-sort.html#nested-sorting

File: config/initializers/tire.rb [hunk 8]
```diff
@@ -1,3 +1,2 @@
-#Tire.configure { logger STDERR, level: 'debug' }
-PositionOpening.create_search_index unless PositionOpening.search_index.exists?
-Geoname.create_search_index unless Geoname.search_index.exists?
\ No newline at end of file
+PositionOpening.create_search_index unless PositionOpening.search_index_exists?
```
Comment (nickmarden) [inline config/initializers/tire.rb]: Consider renaming this initializer since the `tire` gem is no longer present?

**Review by nickmarden — CHANGES_REQUESTED** [review]

File: config/initializers/elasticsearch.rb [hunk 9]
```diff
@@ -1,5 +1,25 @@
...
+  Rails.application.config.elasticsearch_client = Elasticsearch::Client.new(
+    host: [{
+      url: config['url'],
+      user: config['username'],
+      password: config['password']
+    }]
+  )
+else
+  Rails.application.config.elasticsearch_client = Elasticsearch::Client.new(
+    host: config['host'],
```
Comment (nickmarden) [inline config/initializers/elasticsearch.rb]: I think you're assuming too much about whether I have a username or password in different environments. If there is a username/password, for the given environment, pass it. If not, don't.

Similarly, you're assuming a lot about whether there are multiple hosts in production or development/test. Have the code work the same in both cases unless there is a good reason not to.

File: config/initializers/elasticsearch.rb [hunk 10]
```diff
@@ -1,5 +1,25 @@
 module Elasticsearch; end
 
-es_config = Rails.application.config.elasticsearch_config
+Rails.application.config.elasticsearch = (YAML.load_file("#{Rails.root}/config/elasticsearch.yml") || {})[Rails.env]
```
Comment (nickmarden) [inline config/initializers/elasticsearch.rb]: Can you use `config_for` here?

File: app/classes/search_definition.rb [hunk 11]
```diff
@@ -0,0 +1,40 @@
+class SearchDefinition
```
Comment (nickmarden) [inline app/classes/search_definition.rb]: Feels like what your `must`, `should`, etc. queries would be easier with https://github.com/rails/jbuilder?

Comment (MothOnMars) [reply app/classes/search_definition.rb]: Or better yet, the ES DSL: https://github.com/elastic/elasticsearch-ruby/tree/master/elasticsearch-dsl

File: spec/models/geoname_spec.rb [hunk 12]
```diff
@@ -2,38 +2,39 @@
...
-    Geoname.delete_search_index if Geoname.search_index.exists?
+    Geoname.delete_search_index if Geoname.search_index_exists?
     Geoname.create_search_index
   end
 
   describe '.geocode(options)' do
 
     describe 'basic location/state lookup' do
-      before do
+      before(:each) do
```
Comment (nickmarden) [inline spec/models/geoname_spec.rb]: isn't `before(:each)` unnecessary? I think it meant the same as just `before`

File: app/models/geoname.rb [hunk 13]
```diff
@@ -1,75 +1,112 @@
...
+              { match: { location: { query: options[:location], operator: 'AND' } } },
+              { term: { state: options[:state] } }
+            ]
+          }
+        }
+      }
+
+      search_definition[:size] = options[:size]
+
+      Geoname.search(search_definition, index: INDEX_NAME)
```
Comment (nickmarden) [inline app/models/geoname.rb]: Is it possible to use `Elasticsearch::DSL::Search` to be consistent with `PositionOpening`?

File: app/models/geoname.rb [hunk 14]
```diff
@@ -1,75 +1,112 @@
...
     end
 
     def delete_search_index
-      search_index.delete
+      client.indices.delete index: INDEX_NAME rescue nil
```
Comment (nickmarden) [inline app/models/geoname.rb]: I think you're rescuing `nil` here because `client.indices` might not exist for `INDEX_NAME`. In that case could you either check first (`client.indices.delete(index: INDEX_NAME) if search_index_exists?`) or rescue a more specific error here? I don't want to drop other exceptions on the floor unnecessarily.

**Review by nickmarden — CHANGES_REQUESTED** [review]

File: spec/models/geoname_spec.rb [hunk 15]
```diff
@@ -14,26 +14,27 @@
       end
 
       it 'should return the lat/lon hash of the place' do
-        expect(Geoname.geocode(location: "Someplace", state: 'XY')).to eq({lat: 12.34, lon: -123.45})
+        Geoname.geocode(location: "Someplace", state: 'XY')
```
Comment (nickmarden) [inline spec/models/geoname_spec.rb]: Why is this line necessary?

File: spec/models/geoname_spec.rb [hunk 16]
```diff
@@ -14,26 +14,27 @@
...
-            geonames << {type: 'geoname', location: "#{synonym} City", state: 'CA', geo: {lat: rand * 180, lon: rand * 180}}
+            geonames << {type: 'geoname', location: "#{synonym} City", state: 'CA', geo: {lat: rand * 90, lon: rand * 180}}
           end
         end
         Geoname.import geonames
       end
 
       it 'should find the matches' do
         @first_synonyms.each do |synonym|
+          Geoname.geocode(location: "#{synonym} City", state: 'CA')
```
Comment (nickmarden) [inline spec/models/geoname_spec.rb]: Same question here.

File: spec/models/geoname_spec.rb [hunk 17]
```diff
@@ -45,11 +46,12 @@
   describe '.import(geonames)' do
     it 'should set the document ID' do
       Geoname.import [{type: 'geoname', location: "Someplace", state: 'XY', geo: {lat: 12.34, lon: -123.45}}]
-      Geoname.import [{type: 'geoname', location: "Someplace", state: 'XY', geo: {lat: 92.34, lon: 23.45}}]
+      Geoname.import [{type: 'geoname', location: "Someplace", state: 'XY', geo: {lat: 82.34, lon: 23.45}}]
+      Geoname.search_for(location: 'Someplace', state: 'XY', size: 2)
```
Comment (nickmarden) [inline spec/models/geoname_spec.rb]: Same question here.

File: app/models/position_opening.rb [hunk 18]
```diff
@@ -1,134 +1,224 @@
...
-              analysis: {
-                analyzer: { custom_analyzer: { type: 'custom', tokenizer: 'whitespace', filter: %w(standard lowercase synonym snowball) } },
-                filter: { synonym: { type: 'synonym', synonyms: SYNONYMS } }
-              }
-            }
-          },
-          mappings: {
-            position_opening: {
-              _timestamp: { enabled: true },
-              _ttl: { enabled: true },
```
Comment (nickmarden) [inline app/models/position_opening.rb:25]: I think there is something important going on here regarding the special `_ttl` field being changed to a "regular" field called `ttl`. The Elasticsearch cluster will no longer get rid of documents whose `ttl` duration has passed.

Were you aware of this issue @yih-en, and if so, did you have a plan to replace this functionality?

File: app/models/position_opening.rb [hunk 19]
```diff
@@ -141,28 +231,40 @@ def search_for(options = {})
...
-        import position_openings do |docs|
-          docs.each do |doc|
-            doc[:id] = "#{doc[:source]}:#{doc[:external_id]}"
-            doc[:locations].each do |loc|
-              normalized_city = loc[:city].sub(' Metro Area', '').sub(/, .*$/, '')
-              lat_lon_hash = Geoname.geocode(location: normalized_city, state: loc[:state])
-              loc[:geo] = lat_lon_hash if lat_lon_hash.present?
-            end if doc[:locations].present?
+      position_openings.each do |opening|
+        data = opening.except(:_timestamp, :_ttl).each_with_object({}) do |(key, value), data|
```
Comment (nickmarden) [inline app/models/position_opening.rb]: If `_timestamp` and `_ttl` are no longer special fields in Elasticsearch (they aren't), then I'd like the importer to no longer expect them to have the leading underscores in the field names. It's just unnecessarily mysterious.

File: app/models/position_opening.rb [hunk 20]
```diff
@@ -252,12 +252,15 @@ def import(position_openings)
...
 
+        ts = opening[:timestamp].blank? ? DateTime.current : opening[:timestamp]
+
         client.index(
           index: INDEX_NAME,
           type: 'position_opening',
           id: "#{opening[:source]}:#{opening[:external_id]}",
           body: data.merge!({
-            timestamp: opening[:_timestamp].blank? ? DateTime.current : opening[:_timestamp],
+            timestamp: ts,
```
Comment (nickmarden) [inline app/models/position_opening.rb]: I even think this code could be moved up to https://github.com/GSA/jobs_api/pull/17/commits/f522c3aee176a67e93a3491a42f1a77bb7650cb6#diff-681110cebeffad0ee72c994adfb4c087R244 as part of a case statement.

File: config/schedule.rb [hunk 21]
```diff
@@ -0,0 +1,14 @@
+require "active_support"
+require "active_support/time"
+
+Time.zone = "Eastern Time (US & Canada)"
+
```
Comment (yih-en) [inline config/schedule.rb:5]: @nickmarden, kindly advise on the timezone for cronjob

File: config/initializers/elasticsearch.rb [hunk 22]
```diff
@@ -1,7 +1,16 @@
 module Elasticsearch; end
 
-es_config = (YAML.load_file("#{Rails.root}/config/elasticsearch.yml") || {})[Rails.env]
+config = Rails.application.config.elasticsearch
 
-Tire::Configuration.url(es_config['url']) if es_config && es_config['url'].present?
+Elasticsearch::INDEX_NAME = config && config['index_name'].present? ? config['index_name'].freeze : "#{Rails.env}:jobs".freeze
 
-Elasticsearch::INDEX_NAME = es_config && es_config['index_name'].present? ? es_config['index_name'].freeze : "#{Rails.env}:jobs".freeze
+Rails.application.config.elasticsearch_client = Elasticsearch::Client.new(
```
Comment (viclim) [inline config/initializers/elasticsearch.rb]: Why store instance of `Elasticsearch::Client` in rails config?
Also you can do access the configurations with `Rails.configuration.elasticsearch`

File: app/models/position_opening.rb (hunk 18 above)
Comment (yih-en) [reply app/models/position_opening.rb:25]: We can set up a cron job to delete the expired/invalid job posts: [See here](https://github.com/GSA/jobs_api/pull/17/files#diff-681110cebeffad0ee72c994adfb4c087R293)
